### PR TITLE
mindforger: 1.52.0 -> 2.0.0

### PR DESCRIPTION
**Pull request by lopter**
Pull MindForger's [fork][cmark-fork] of [cmark-gfm][cmark] alongside Mindforger and build it in `preBuild`. I am not sure to understand how mindforger could be built before that.

The patches that were applied have all been upstreamed and could be removed.

A patch for the Qt project file to use pkg-config to find hunspell was added.

[cmark]: https://github.com/github/cmark-gfm
[cmark-fork]: https://github.com/dvorka/cmark

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lopter** [comment]
Patch is the same for NixOS 23.11, except `wrapGAppsHook` is used instead of `wrapGAppsHook3`.

**Comment by JohnAZoidberg** [comment]
I updated the PR title to match our guidelines.

File: pkgs/applications/editors/mindforger/default.nix [hunk 1]
```diff
@@ -1,37 +1,56 @@
...
+  preBuild = ''(
+      mkdir deps/cmark-gfm/build &&
+      cd deps/cmark-gfm/build &&
+      cmake -DCMARK_TESTS=OFF -DCMARK_SHARED=OFF .. &&
+      cmake --build . --parallel
+  )'';
 
   doCheck = true;
 
   patches = [
```
Comment (FliegendeWurst) [inline pkgs/applications/editors/mindforger/default.nix:52]: Please set meta.platforms = linux or fix
```
       > ranlib -s libmindforger.a
       > ranlib: file: libmindforger.a(link.o) has no symbols
       > ranlib: file: libmindforger.a(naive_autolinking_preprocessor.o) has no symbols
       > make[1]: Leaving directory '/private/tmp/nix-build-mindforger-2.0.0.drv-0/source/lib'
       > cd app/ && ( test -e Makefile || /nix/store/n84f7j30ni38zl56rfih32nc6yp30qcd-qtbase-5.15.12-dev/bin/qmake -o Makefile /private/tmp/nix-build-mindforger-2.0.0.drv-0/source/app/app.pro PREFIX=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0 NIX_OUTPUT_OUT=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0 NIX_OUTPUT_DEV=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0 NIX_OUTPUT_BIN=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0 NIX_OUTPUT_DOC=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0/share/doc/qt-5.15.12 NIX_OUTPUT_QML=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0/lib/qt-5.15.12/qml NIX_OUTPUT_PLUGIN=/nix/store/ma3qya28jjrwpbq04pasw9yl1pw9p2j1-mindforger-2.0.0/lib/qt-5.15.12/plugins CONFIG+=release CONFIG+=nostrip CONFIG+=mfnoccache CONFIG+=mfwebengine ) && make -f Makefile
       > make[1]: Entering directory '/private/tmp/nix-build-mindforger-2.0.0.drv-0/source/app'
       > make[1]: *** No rule to make target '../doc/limbo', needed by 'mindforger.app/Contents/Resources/mindforger-repository/limbo'.  Stop.
       > make[1]: Leaving directory '/private/tmp/nix-build-mindforger-2.0.0.drv-0/source/app'
       > make: *** [Makefile:76: sub-app-make_first] Error 2
```

Comment (lopter) [reply pkgs/applications/editors/mindforger/default.nix:52]: On which platform are you getting that? Darwin?

Comment (lopter) [reply pkgs/applications/editors/mindforger/default.nix:52]: What seems to happen is that the Mindforger packages the documentation on MacOS [1], and it's pulled to the doc folder as a git submodule [2].

[1]: https://github.com/dvorka/mindforger/blob/de1da1d50582d3a34aa977bbd689505c3992a669/app/app.pro#L494
[2] https://github.com/dvorka/mindforger/blob/de1da1d50582d3a34aa977bbd689505c3992a669/.gitmodules#L1

Would mind trying something like:

```diff
diff --git a/pkgs/applications/editors/mindforger/default.nix b/pkgs/applications/editors/mindforger/default.nix
index 722a425ca27b..bbda7a063c77 100644
--- a/pkgs/applications/editors/mindforger/default.nix
+++ b/pkgs/applications/editors/mindforger/default.nix
@@ -26,6 +26,12 @@ let
       rev = "4ca8688093228c599432a87d7bd945804c573d51";
       sha256 = "sha256-0WiG8aot8mc0h1BKPgC924UKQrgunZvKKBy9bD7nhoQ=";
     };
+    mindforger-repository = {
+      owner = "dvorka";
+      repo = "mindforger-repository";
+      rev = "4ca8688093228c599432a87d7bd945804c573d51";
+      sha256 = "";
+    };
   };
 in stdenv.mkDerivation {
   pname = "mindforger";
@@ -37,8 +43,13 @@ in stdenv.mkDerivation {
   buildInputs = [ curl hunspell qtbase qtwebengine ];

   # Disable the cmake hook (so we don't try to build MindForger with it), and
-  # build MindForger's internal fork of cmark-gfm ahead of MindForger itself:
-  postUnpack = "cp -TR ${srcs.cmark-gfm} $sourceRoot/deps/cmark-gfm";
+  # build MindForger's internal fork of cmark-gfm ahead of MindForger itself.
+  #
+  # Moreover unpack the docs that are needed for the MacOS build.
+  postUnpack = ''
+    cp -TR ${srcs.cmark-gfm} $sourceRoot/deps/cmark-gfm
+    cp -TR ${srcs.mindforger-repository} $sourceRoot/doc
+  '';
   dontUseCmakeConfigure = true;
   preBuild = ''(
       mkdir deps/cmark-gfm/build &&
```

On top of this PR?

Comment (FliegendeWurst) [reply pkgs/applications/editors/mindforger/default.nix:52]: Well I can't. I just looked at ofborg's build attempt. (You can push your fix with the right hash to test)

**Comment by lopter** [comment]
Fix string concatenation, thanks @FliegendeWurst.

**Comment by FliegendeWurst** [comment]
@ofborg build mindforger

**Comment by lopter** [comment]
Looks like [there is](https://logs.ofborg.org/?attempt_id=6dd05986-2f22-410c-9315-577ba084bf58&key=nixos%2Fnixpkgs.314128) some error trying to build qtwebengine on Darwin:

```
error: builder for '/nix/store/sbxrgl9hsfjv4wyjslaacnb97ncl4k6j-qtwebengine-5.15.17.drv' failed with exit code 2;
       last 25 log lines:
       > 2 warnings generated.
       > [3118/23860] CXX obj/third_party/harfbuzz-ng/harfbuzz_source/hb-ft.o[K[Kalize.o
       > FAILED: obj/third_party/harfbuzz-ng/harfbuzz_source/hb-ft.o
       > /nix/store/bhqyl2380k1nhwxw58ki7k9rdgnks1bd-clang-wrapper-16.0.6/bin/clang++ -MMD -MF obj/third_party/harfbuzz-ng/harfbuzz_source/hb-ft.o.d -DHAVE_OT -DHAVE_FREETYPE -DHAVE_ICU -DHAVE_ICU_BUILTIN -DHAVE_INTEL_ATOMIC_PRIMITIVES -DHAVE_ROUNDF -DHB_NO_MMAP -DHB_NO_RESOURCE_FORK -DHB_NO_SUBSET_LAYOUT -DHB_NO_FALLBACK_SHAPE -DHB_NO_UCD -DHB_NO_METRICS -DHB_NO_WIN1256 -DU_DISABLE_VERSION_SUFFIX=0 -DOFFICIAL_BUILD -DTOOLKIT_QT -D_LIBCPP_HAS_NO_ALIGNED_ALLOCATION -DCR_XCODE_VERSION=0010 -D__ASSERT_MACROS_DEFINE_VERSIONS_WITHOUT_UNDERSCORES=0 -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DFT_CONFIG_MODULES_H=\"freetype-custom/freetype/config/ftmodule.h\" -DFT_CONFIG_OPTIONS_H=\"freetype-custom/freetype/config/ftoption.h\" -DPDFIUM_REQUIRED_MODULES -DCHROMIUM_RESTRICT_VISIBILITY -DU_USING_ICU_NAMESPACE=0 -DU_ENABLE_DYLOAD=0 -DUSE_CHROMIUM_ICU=1 -DU_ENABLE_TRACING=1 -DU_ENABLE_RESOURCE_TRACING=0 -DU_STATIC_IMPLEMENTATION -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_FILE -DUCHAR_TYPE=uint16_t -Igen -I../../3rdparty/chromium -I../../3rdparty/chromium/third_party/freetype/include -I../../3rdparty/chromium/third_party/freetype/include/freetype-custom -I../../3rdparty/chromium/third_party/freetype/src/include -I../../3rdparty/chromium/third_party/harfbuzz-ng/src/src -I../../3rdparty/chromium/third_party/icu/source/common -I../../3rdparty/chromium/third_party/icu/source/i18n -fno-strict-aliasing -fstack-protector -Wno-unknown-attributes -Wno-unknown-pragmas -Wno-unknown-warning-option -fcolor-diagnostics -fmerge-all-constants -arch x86_64 -fno-omit-frame-pointer -g0 -isysroot ../../../../../../../nix/store/4nhcmnam6rp1w46hyfcizxh76rwism2l-apple-sdk-13.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -mmacosx-version-min=10.13 -fvisibility=hidden -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Wall -Wno-unused-variable -Wno-misleading-indentation -Wunguarded-availability -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -Wno-psabi -Wno-ignored-pragma-optimize -Wno-implicit-int-float-conversion -Wno-final-dtor-non-final-class -Wno-builtin-assume-aligned-alignment -Wno-deprecated-copy -Wno-non-c-typedef-for-linkage -Wno-max-tokens -O3 -Wno-parentheses-equality -Wno-tautological-compare -Wno-thread-safety-attributes -Wno-undefined-bool-conversion -Wno-tautological-undefined-compare -std=c++14 -fno-trigraphs -Wno-trigraphs -stdlib=libc++ -fno-exceptions -fno-rtti -fvisibility-inlines-hidden -c ../../3rdparty/chromium/third_party/harfbuzz-ng/src/src/hb-ft.cc -o obj/third_party/harfbuzz-ng/harfbuzz_source/hb-ft.o
       > ../../3rdparty/chromium/third_party/harfbuzz-ng/src/src/hb-ft.cc:759:73: error: cast from 'void (*)(FT_Face)' (aka 'void (*)(FT_FaceRec_ *)') to 'FT_Generic_Finalizer' (aka 'void (*)(void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
       >   if (unlikely (!ft_face->generic.data || ft_face->generic.finalizer != (FT_Generic_Finalizer) hb_ft_face_finalize))
       >                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > ../../3rdparty/chromium/third_party/harfbuzz-ng/src/src/hb.hh:234:46: note: expanded from macro 'unlikely'
       > #define unlikely(expr) (__builtin_expect (!!(expr), 0))
       >                                              ^~~~
       > ../../3rdparty/chromium/third_party/harfbuzz-ng/src/src/hb-ft.cc:765:34: error: cast from 'void (*)(FT_Face)' (aka 'void (*)(FT_FaceRec_ *)') to 'FT_Generic_Finalizer' (aka 'void (*)(void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
       >     ft_face->generic.finalizer = (FT_Generic_Finalizer) hb_ft_face_finalize;
       >                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > ../../3rdparty/chromium/third_party/harfbuzz-ng/src/src/hb-ft.cc:1035:32: error: cast from 'void (*)(FT_Face)' (aka 'void (*)(FT_FaceRec_ *)') to 'FT_Generic_Finalizer' (aka 'void (*)(void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
       >   ft_face->generic.finalizer = (FT_Generic_Finalizer) _release_blob;
       >                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > 3 errors generated.
       > ninja: build stopped: subcommand failed.
       > make[3]: *** [Makefile.gn_run:390: run_ninja] Error 1
       > make[3]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.15.17.drv-0/source/src/core'
       > make[2]: *** [Makefile:89: sub-gn_run-pro-make_first] Error 2
       > make[2]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.15.17.drv-0/source/src/core'
       > make[1]: *** [Makefile:87: sub-core-make_first] Error 2
       > make[1]: Leaving directory '/private/tmp/nix-build-qtwebengine-5.15.17.drv-0/source/src'
       > make: *** [Makefile:56: sub-src-make_first] Error 2
       For full logs, run 'nix log /nix/store/sbxrgl9hsfjv4wyjslaacnb97ncl4k6j-qtwebengine-5.15.17.drv'.
error: 1 dependencies of derivation '/nix/store/9vzv23lq4n3rk2957h9hlyy6yvbg8k9h-mindforger-2.0.0.drv' failed to build
```

Seems to originate from some version of harfbuzz vendored with Bazel…

Following it through Github got me here : https://github.com/qt/qtwebengine-chromium/tree/4140e17e7f2935fcdd4816c4050e4d2eebda2be9/chromium/third_party/harfbuzz-ng

**Review by FliegendeWurst — COMMENTED** [review]
Would be good to squash the commits

**Comment by lopter** [comment]
- [x] Squash commits

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**